### ISSUE TITLE
Models treating text fields as booleans for filtering

### ISSUE THREAD

**Issue by megansmcguire** [metabase/metabase#47639]
### Describe the bug

<img width="314" alt="Screenshot 2024-09-04 at 5 31 16 PM" src="https://github.com/user-attachments/assets/f1326078-ec5d-456e-bc78-f12ee6c4d827">

Since updating to Metabase 50 yesterday, I have a number of text/categorical fields in models that only want to filter as booleans (see example above), even after changing the table metadata and the column metadata in the model. This is leaving me unable to filter these fields at all. This happens both when I create a new model or I edit an existing model based on this table.

### To Reproduce

Unable to reproduce with sample database

### Expected behavior

Should align with the field values, not a boolean

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/128.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "metabase-info": {
    "databases": [
      "snowflake",
      "postgres",
      "mysql",
      "h2"
    ],
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2024-09-03",
      "tag": "v0.50.23.2",
      "hash": "95f1c3b"
    },
    "settings": {
      "report-timezone": "UTC"
    },
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "13.12"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.3"
      }
    }
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.24+8",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.24",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.24+8",
    "os.name": "Linux",
    "os.version": "5.10.223-212.873.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  }
}
```

### Severity

Pretty severe; several models are not working because of this

**Comment by ranquild** [comment]
Likely caused by https://github.com/metabase/metabase/pull/45674 and Snowflake's `VARIANT` type.

**Comment by ranquild** [comment]
@megansmcguire meanwhile you could click "Custom Expression" when filtering and use custom expressions like `contains([Priority], "P1")`

**Comment by megansmcguire** [comment]
TY, should've thought of that! Will help for now

**Comment by ranquild** [comment]
The current state of things:
1. For `VARIANT` columns, we will render a widget that allows to use only `Is empty` and `Not empty` operators. You can use `Custom expression` and use any expression as long as the database can execute it.
2. If you want to get typed columns, you would need to create a SQL model/question and explicitly cast the column. If you cast it to a number, you would get a widget with numeric operators like `Greater than` etc.

**Comment by iethree** [comment]
🚀 This should also be released by [v0.51](https://github.com/metabase/metabase/milestone/231)
